BUG: Fix #4476 by adding datetime64 and timedelta64 types

gerritholl opened the pull request:
This commit fixes bug #4476 by adding the codes for the datetime64 and
timedelta64 types to the `default_filler` dictionary in numpy.ma.core,
used by `default_fill_value`.  Also adapt checking in the
`default_fill_value` to include code for timedelta64, not only datetime64.

gerritholl commented:
This is a second attempt to provide the pull request in the correct format, fixing technical git issues that I got wrong in #4477.

charris commented:
LGTM. Needs a test, probably in `TestFillingValues` in `numpy/ma/tests/test_core.py`. Look at how the other tests are done.

You managed to recover ;) Any problems?

gerritholl commented:
I added a testcase.  Should I merge the commit for fixing the problem and the commit for adding the testcase to a single commit?

gerritholl commented:
I did notice one small thing; the representation for `numpy.datetime64('NaT')` does not indicate the timecode, but `numpy.ma.testutils.assert_equal` does (fortunately) pay attention to this.  Consequently, if the timecode is wrong, the testing system will report:

```
Items are not equal:
 ACTUAL: numpy.datetime64('NaT')
 DESIRED: numpy.datetime64('NaT')
```

which is potentially going to confuse somebody in the future.  But that's a different issue (should `datetime64.__repr__` show the time code?).

gerritholl commented:
It does appear I have messed up again.  I will try again to fix it.

charris commented:
Just say no to merge, use rebase ;) That said, the only reason to rebase is 1) Clean up history, rewrite commit messages, etc. 2) PR is not mergable into master.

No need to squash the two commits, but the second should begin `TST: `.

charris commented:
It's easy to recover from a bad merge, just `git reset --hard ORIG_HEAD`.

gerritholl commented:
It seems the commit message and diff are correct now, I don't know why it says the commit is authored by JDWarner though...

Commit:
BUG: Fix #4476 by adding datetime64 and timedelta64 types

This commit fixes bug #4476 by adding the codes for the datetime64 and
timedelta64 types to the `default_filler` dictionary in numpy.ma.core,
used by `default_fill_value`.  Also adapt checking in the
`default_fill_value` to include code for timedelta64, not only datetime64.

Commit:
TST: Add testcase for the fix to bug #4476.

Add a testcase `test_fillvalue_datetime_timedelta` to class
`TestFillingValues` for the fix to bug #4476.  See commit
216fd17 and pull request #5455.

charris commented:
Hmm, not sure how you got there, branch looks OK `git log --graph`. You can use `git commit --amend` to change the author. The lines of the commit message should also be at most 72 characters long.

gerritholl commented:
Ok.  I hope it's all OK now.

gerritholl commented:
Do I need to do anything to keep this "up to date" as this pull request stands while the upstream/numpy main branch moves on?

charris commented:
Nope, not unless the merge button turns grey and the PR cannot be merged. That is unlikely unless someone edits the same files in almost the same places.

charris commented:
Merged, thanks @gerritholl .
